xla/service/gpu/transforms/dynamic_slice_fusion_rewriter.cc

```diff
@@ -152,99 +158,439 @@ bool IsAlignedSlice(const HloInstruction* slice) {
-//   c_trip_count = s32[] constant(16)
-//   add = s32[] add(gte, c_trip_count) select = s32[] select(compare, add, gte)
-// clang-format on
+// Function looks for while backend config. If this config is present, it
+// returns the value of trip count, otherwise it runs the while loop analysis to
+// compute trip count. `whileop` must be a while operaton. Returns
+// `std::nullopt` if it cannot figure out the trip count.
+std::optional<int64_t> GetWhileLoopTripCount(HloInstruction* whileop) {
+  CHECK(whileop->opcode() == HloOpcode::kWhile);
+  auto backend_config = whileop->backend_config<WhileLoopBackendConfig>();
+  if (!backend_config.ok() || !backend_config.value().has_known_trip_count()) {
+    VLOG(4) << "Backend config not ok. Computing while loop trip count for "
+            << whileop->name();
+    return ComputeWhileLoopTripCount(whileop);
+  }
+  int trip_count = backend_config.value().known_trip_count().n();
+  VLOG(4) << "Found trip count in backend config for " << whileop->name()
+          << ": " << trip_count;
+  return trip_count;
+}
 
-bool IsLoopIterationNumber(const HloInstruction& offset) {
-  const HloComputation* parent = offset.parent();
-  if (!parent->IsWhileBodyComputation()) return false;
-
-  // Scan loops trip count must be known at compile time as it iterates over the
-  // leading dimension of the statically shaped input.
-  const HloInstruction* while_instr = parent->WhileCallInstruction();
-  auto config = while_instr->backend_config<xla::WhileLoopBackendConfig>();
-  if (!config.ok() || !config->has_known_trip_count()) return false;
-  int32_t trip_count = config->known_trip_count().n();
-
-  // First lets check the offset computation pattern
-  if (!Match(&offset, m::Select(m::Lt(m::GetTupleElement(m::Parameter(0)),
-                                      m::ConstantScalar<int32_t>(0)),
-                                m::Add(m::GetTupleElement(m::Parameter(0)),
-                                       m::ConstantScalar(trip_count)),
-                                m::GetTupleElement(m::Parameter())))) {
+// Given an operation `op`, an empty module `module` and an induction variable
+// (input) `indvar`, this function traverses the computation in reverse post
+// order and tries to build a computation of the form: `op = f(indvar)`. If it
+// is successful in making such a computation, then this computation is inserted
+// as an entry computation in the empty module. Now, this module can be invoked
+// with multiple values of the `indvar` to get the value of the operation at
+// that value of `indvar`. This is used to get two computations: `offset = f(i)`
+// and `updated_i = g(i)`. If such a computation cannot be created within the
+// module (maybe the offset depends on partition-id or something else), then it
+// returns false. If it is successful, it returns true.
+bool GetOperationModule(const HloInstruction* op, HloModule* module,
```

Comment:
Can this method just return `std::unique_ptr<HloModule`  instead of bool? You can let it return nullptr in case it doesn't succeed.

Reply:
Removed this function as suggested.